linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/SJSGen.scala

```diff
@@ -360,11 +377,37 @@ private[emitter] final class SJSGen(
     }
   }
 
+  def genNewArray(arrayTypeRef: ArrayTypeRef, lengths: List[Tree])(
+      implicit moduleContext: ModuleContext, globalKnowledge: GlobalKnowledge,
+      pos: Position): Tree = {
+    lengths match {
+      case length :: Nil =>
+        New(genArrayConstrOf(arrayTypeRef), length :: Nil)
+      case _ =>
+        genCallHelper("newArrayObject", genClassDataOf(arrayTypeRef),
+            ArrayConstr(lengths))
+    }
+  }
+
   def genArrayValue(arrayTypeRef: ArrayTypeRef, elems: List[Tree])(
       implicit moduleContext: ModuleContext, globalKnowledge: GlobalKnowledge,
       pos: Position): Tree = {
-    genCallHelper("makeNativeArrayWrapper", genClassDataOf(arrayTypeRef),
```

Comment:
`ResolvedApply` from #4337 takes this direction. I think that would be a nicer interface than the stringly typed one we have now.

Reply:
I have revisited that a bit. I still think that a refactoring to separate the two call helper interfaces should be done as a follow-up. However, it turns out there was a good reason in this commit to collapse the 3 helpers into a single one, because that allows to generate much better code for `ArrayBuilder.result`. When we statically know the element type, and it's one of the specialized types, we can emit a direct call like `new $ac_I(elems)`, which is basically optimal.

I have updated the PR with that change.